### DISCUSSION TITLE
CHAPTER5 - QnA

### DISCUSSION THREAD

**Discussion by JoisFe** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#59] [CHAPTER 6 스트림으로 데이터 수집 - Q&A]
# CHAPTER5 - QnA

### Q1) 스트림 메서드 중 collect와 reduce를 보면 스트림의 요소를 처리하여 데이터를 수집하는 작업에 사용 (reducing 연산) 되는 공통점이 있는데 이 두 메서드의 차이는 ? (각 메서드의 목적, 동작 방식을 기준으로)

**Comment by corock** [comment]
아래 `Stream` 인터페이스 내 추상 메서드 정의를 보면,
각각 `BinaryOperator` 타입과 `Collector` 타입을 받는다는 점에서 큰 차이를 보이고 있습니다.

의미 그대로 `reduce` 메서드는 **이항 연산**을 통해 누적합이나 최솟값, 최댓값 등 스트림 내의 요소로 딱 떨어지는 값을 도출할 때 사용하고,
이에 반해 `collect` 메서드는 모으다라는 의미처럼 스트림 내의 요소를 수집해서 다른 컬렉션 타입으로 제공할 때 주로 사용하지 않을까 싶습니다.

```java
public interface Stream<T> extends BaseStream<T, Stream<T>> {

    // (...)

    T reduce(T identity, BinaryOperator<T> accumulator);

    // (...)

    <R, A> R collect(Collector<? super T, A, R> collector);

    // (...)

}
```

**Comment by umr-x** [comment]
좋은 설명에 덧붙여서, 간단한 요약과 실용적인 예시를 공유하고자 합니다.

* **reduce()**: 요소들을 하나의 값으로 '결합'하고 싶을 때 사용합니다 (합계, 최솟값, 최댓값 계산 등).
* **collect()**: 요소들을 다른 구조로 '수집'하고 싶을 때 사용합니다 (리스트를 Set이나 Map으로 변환 등).

**간단한 예시:**
```java
// reduce: 숫자들의 합계
int sum = numbers.stream().reduce(0, (a, b) -> a + b);

// collect: 리스트로 수집
List<String> list = names.stream().collect(Collectors.toList());
